tests/contrib/operators/test_gcp_function_operator.py

```diff
@@ -0,0 +1,624 @@
+# or more contributor license agreements.  See the NOTICE file
+# distributed with this work for additional information
+# regarding copyright ownership.  The ASF licenses this file
+# to you under the Apache License, Version 2.0 (the
+# "License"); you may not use this file except in compliance
+# with the License.  You may obtain a copy of the License at
+#
+#   http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing,
+# software distributed under the License is distributed on an
+# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+# KIND, either express or implied.  See the License for the
+# specific language governing permissions and limitations
+# under the License.
+
+import unittest
+
+from googleapiclient.errors import HttpError
+from parameterized import parameterized
+
+from airflow.contrib.operators.gcp_function_operator import \
+    GcfFunctionDeployOperator, GcfFunctionDeleteOperator, FUNCTION_NAME_PATTERN
+from airflow import AirflowException
+from airflow.version import version
+
+from copy import deepcopy
+
+try:
+    # noinspection PyProtectedMember
+    from unittest import mock
+except ImportError:
+    try:
+        import mock
+    except ImportError:
+        mock = None
+
+EMPTY_CONTENT = ''.encode('utf8')
+MOCK_RESP_404 = type('', (object,), {"status": 404})()
+
+PROJECT_ID = 'test_project_id'
+LOCATION = 'test_region'
+SOURCE_ARCHIVE_URL = 'gs://folder/file.zip'
+ENTRYPOINT = 'helloWorld'
+FUNCTION_NAME = 'projects/{}/locations/{}/functions/{}'.format(PROJECT_ID, LOCATION,
+                                                               ENTRYPOINT)
+RUNTIME = 'nodejs6'
+VALID_RUNTIMES = ['nodejs6', 'nodejs8', 'python37']
+VALID_BODY = {
+    "name": FUNCTION_NAME,
+    "entryPoint": ENTRYPOINT,
+    "runtime": RUNTIME,
+    "httpsTrigger": {},
+    "sourceArchiveUrl": SOURCE_ARCHIVE_URL
+}
+
+
+def _prepare_test_bodies():
+    body_no_name = deepcopy(VALID_BODY)
+    body_no_name.pop('name', None)
+    body_empty_entry_point = deepcopy(VALID_BODY)
+    body_empty_entry_point['entryPoint'] = ''
+    body_empty_runtime = deepcopy(VALID_BODY)
+    body_empty_runtime['runtime'] = ''
+    body_values = [
+        ({}, "The required parameter 'body' is missing"),
+        (body_no_name, "The required body field 'name' is missing"),
+        (body_empty_entry_point,
+         "The body field 'entryPoint' of value '' does not match"),
+        (body_empty_runtime, "The body field 'runtime' of value '' does not match"),
+    ]
+    return body_values
+
+
+class GcfFunctionDeployTest(unittest.TestCase):
+    @parameterized.expand(_prepare_test_bodies())
+    @mock.patch('airflow.contrib.operators.gcp_function_operator.GcfHook')
+    def test_body_empty_or_missing_fields(self, body, message, mock_hook):
+        mock_hook.return_value.upload_function_zip.return_value = 'https://uploadUrl'
+        with self.assertRaises(AirflowException) as cm:
```

Comment:
nit: Not sure what `cm` stands for, maybe call it `e` or `expected`?

Reply:
Yeah. 'e' is better. I have no idea either what cm was for.

Reply:
It stands for "context manager" and it actually is not the exception itself, so I would suggest against changing it to `e`.

Take a look at:
```
with self.assertRaises(SomeException) as cm:
    do_something()

the_exception = cm.exception
self.assertEqual(the_exception.error_code, 3)
```
from https://docs.python.org/2/library/unittest.html

Reply:
OK. Good. I reverted it back. It makes perfect sense.